packages/contracts-bedrock/contracts/EAS/EAS.sol

```diff
@@ -0,0 +1,853 @@
+
+pragma solidity 0.8.19;
+
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import { Semver } from "../universal/Semver.sol";
+
+import { EIP712Verifier } from "./eip712/EIP712Verifier.sol";
+
+import { ISchemaResolver } from "./resolver/ISchemaResolver.sol";
+
+// prettier-ignore
+import {
+    AccessDenied,
+    EMPTY_UID,
+    EIP712Signature,
+    InvalidLength,
+    MAX_GAP,
+    NotFound,
+    NO_EXPIRATION_TIME,
+    uncheckedInc
+} from "./Common.sol";
+
+// prettier-ignore
+import {
+    Attestation,
+    AttestationRequest,
+    AttestationRequestData,
+    DelegatedAttestationRequest,
+    DelegatedRevocationRequest,
+    IEAS,
+    MultiAttestationRequest,
+    MultiDelegatedAttestationRequest,
+    MultiDelegatedRevocationRequest,
+    MultiRevocationRequest,
+    RevocationRequest,
+    RevocationRequestData
+} from "./IEAS.sol";
+
+import { ISchemaRegistry, SchemaRecord } from "./ISchemaRegistry.sol";
+
+struct AttestationsResult {
+    uint256 usedValue; // Total ETH amount that was sent to resolvers.
+    bytes32[] uids; // UIDs of the new attestations.
+}
+
+/**
+ * @title EAS - Ethereum Attestation Service
+ */
+contract EAS is IEAS, Semver, Initializable, EIP712Verifier {
+    using Address for address payable;
+
+    error AlreadyRevoked();
+    error AlreadyRevokedOffchain();
+    error AlreadyTimestamped();
+    error InsufficientValue();
+    error InvalidAttestation();
+    error InvalidAttestations();
+    error InvalidExpirationTime();
+    error InvalidOffset();
+    error InvalidRegistry();
+    error InvalidRevocation();
+    error InvalidRevocations();
+    error InvalidSchema();
+    error InvalidVerifier();
+    error Irrevocable();
+    error NotPayable();
+    error WrongSchema();
+
+    // The global schema registry.
+    ISchemaRegistry private immutable _schemaRegistry;
+
+    // The global mapping between attestations and their UIDs.
+    mapping(bytes32 uid => Attestation attestation) private _db;
+
+    // The global mapping between data and their timestamps.
+    mapping(bytes32 data => uint64 timestamp) private _timestamps;
+
+    // The global mapping between data and their revocation timestamps.
+    mapping(address revoker => mapping(bytes32 data => uint64 timestamp)) private _revocationsOffchain;
+
+    // Upgrade forward-compatibility storage gap
+    uint256[MAX_GAP - 3] private __gap;
+
+    /**
+     * @dev Creates a new EAS instance.
+     *
+     * @param registry The address of the global schema registry.
+     */
+    constructor(ISchemaRegistry registry) Semver(1, 0, 0) EIP712Verifier("EAS", "1.0.0") {
+        if (address(registry) == address(0)) {
+            revert InvalidRegistry();
+        }
+
+        _schemaRegistry = registry;
+    }
+
+    /**
+     * @dev Initializes the contract and its parents.
+     */
+    function initialize() external initializer {
```

Comment:
Why does this need an initialize function?

Reply:
You should always call initialize in the constructor so that the impl cannot be initialized by an adversary. There is no storage set in the constructor unless something is set inside of the eip712 verifier but then the upgradable version is needed

Reply:
It looks like the eip712 verifier has no storage based on https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.7.3/contracts/utils/cryptography/draft-EIP712.sol

Would like to confirm with the following command:

```
forge inspect EAS storageLayout
```

Reply:
Yes, the initializer isn't really needed here (since there is no storage initialization, neither in its constructor nor in one of its parents). We are following a convention that every implementation contract should have `initialize` in its interface, but perhaps we should just remove it from EAS and SchemaRegistry. Furthermore, since there are no assignment of privileges or storage initialization in its initializer - initializing it by an adversary doesn't mean anything (although, it's very common for people to report this is a super-duper vulnerability via Immunifi 😅).

I'll remove this code for now.

Reply:
I've removed the initializers. Please take a look.

Reply:
We should def remove `initialize` if its a noop, its just a maintenance burden. You are right that not calling `initialize` in the constructor in this case doesn't cause security issues but its best practice in my opinion to always call `initialize` in the constructor in cases where `initialize` is required